### PR TITLE
feat(pdk): add lmdb metrics

### PR DESCRIPTION
**Pull request by oowl**
### Summary
add lmdb metrics to prometheus and statsd, also extend PDK `kong.node.get_memory_stats()`  lmdb capabilities. 

sister PR : https://github.com/Kong/lua-resty-lmdb/pull/25

### Checklist

- [x] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

FTI-4749

### CODE REVIEW

File: kong/pdk/node.lua
Comment (chronolaw) [inline kong/pdk/node.lua:239]: blank line before `else`

File: kong/plugins/statsd/log.lua
Comment (chronolaw) [inline kong/plugins/statsd/log.lua:315]: blank line before else

File: kong/plugins/statsd/log.lua
Comment (chronolaw) [inline kong/plugins/statsd/log.lua:323]: blank line before else

File: kong/pdk/node.lua
Comment (chronolaw) [inline kong/pdk/node.lua:240]: we should add fields to `lmdb_info` and set to `res.lmdb` finally.

File: kong/plugins/statsd/log.lua
Comment (chronolaw) [inline kong/plugins/statsd/log.lua:19]: redundant space before `=`

**Comment by chronolaw** [comment]
change log needed.

**Comment by chronolaw** [comment]
This PR should be draft until https://github.com/Kong/lua-resty-lmdb/pull/25 is merged

File: kong/pdk/node.lua
Comment (hbagdi) [inline kong/pdk/node.lua:173]: Is this the number of current readers? If yes, would current_readers make more sense?

File: kong/pdk/node.lua
Comment (hbagdi) [inline kong/pdk/node.lua:167]: What is the difference between map_size and max_map_size from the user's perspective?
Can the two be different? Should the user be aware of both values?

File: kong/plugins/prometheus/exporter.lua
Comment (hbagdi) [inline kong/plugins/prometheus/exporter.lua:145]:
```suggestion
    memory_stats.lmdb = prometheus:gauge("memory_lmdb_used_bytes",
```

File: kong/plugins/prometheus/exporter.lua
Comment (hbagdi) [inline kong/plugins/prometheus/exporter.lua:146]:
```suggestion
                                         "Used bytes in LMDB",
```

File: kong/plugins/prometheus/exporter.lua
Comment (hbagdi) [inline kong/plugins/prometheus/exporter.lua:149]:
```suggestion
    memory_stats.lmdb_capacity = prometheus:gauge("memory_lmdb_total_bytes",
```

File: kong/plugins/prometheus/exporter.lua
Comment (hbagdi) [inline kong/plugins/prometheus/exporter.lua:151]: The "kong_subssytem" label is perhaps not required since there is a single LMDB map across HTTP and Stream subsystems.

File: kong/pdk/node.lua
Comment (dndx) [reply kong/pdk/node.lua:167]: Also I have some concerns of the string representation of the size. Will this be difficult to consume in an automated fashion? Should the conversion to text representation done by API instead of the PDK?

Remember PDK is a programmable interface, and integer values seems to be the easiest to work with.

File: .requirements
Comment (chronolaw) [inline .requirements:9]: bump to the formal version before merging.

File: kong/pdk/node.lua
Comment (chronolaw) [inline kong/pdk/node.lua:127]:
```suggestion
  --   -- if the `kong` uses dbless mode, the following will be present:
```

File: kong/plugins/statsd/log.lua
Comment (hanshuebner) [inline kong/plugins/statsd/log.lua:19]: Don't localize this variable.  It makes it harder to recognize what's going on at the call site.

**Comment by hbagdi** [comment]
Please create a documentation PR and do not merge this until the docs PR has an approval.
I really like the fact that we have more metrics coming out but I want to make sure that these metrics are usable from the perspective of our users.

**Comment by attenuation** [comment]
block by https://github.com/Kong/lua-resty-lmdb/pull/25, I am waiting for lua-resty-lmdb side merge, when it gets merged, I will continue to push this PR. cc @hbagdi

**Comment by attenuation** [comment]
waiting for `lua-resty-lmdb` side version bump.

File: .requirements
Comment (chronolaw) [reply .requirements:9]: Now we should use `master` for testing.

File: kong/pdk/node.lua
Comment (chronolaw) [inline kong/pdk/node.lua:133]: Now `max_pages` is gone.

File: kong/plugins/statsd/log.lua
Comment (attenuation) [reply kong/plugins/statsd/log.lua:19]: localize related problem has been discussed in another PR, So I choose to ignore this comment.

File: kong/pdk/node.lua
Comment (attenuation) [reply kong/pdk/node.lua:166]: I delete the `max_map_size` entry.

Comment (attenuation) [reply kong/pdk/node.lua:166]: @dndx  this PDK function has unit arguments that mean the caller can control the output data unit. This behavior was trying to follow this function's existing programming interface. https://github.com/Kong/kong/blob/master/kong/pdk/node.lua#L151

File: kong/pdk/node.lua
Comment (chronolaw) [inline kong/pdk/node.lua:134]:
```suggestion
  --    used_size: "0.02 MiB",
  --    last_used_page: 6,
  --    last_txnid: 2,
  --    max_readers: 126,
  --    current_readers: 16
```

File: kong/pdk/node.lua
Comment (chronolaw) [inline kong/pdk/node.lua:169]:
```suggestion
  --    used_size: "20.48 KB",
  --    last_used_page: 6,
  --    last_txnid: 2,
  --    max_readers: 126,
  --    current_readers: 16
```

File: kong/pdk/node.lua
Comment (chronolaw) [inline kong/pdk/node.lua:260]: Shoud we use `self.table.new` as above?

File: kong/pdk/node.lua
Comment (chronolaw) [inline kong/pdk/node.lua:22]: We have the different name `get_env_info` in statsd, should we give them the same name?

File: kong/plugins/statsd/log.lua
Comment (hanshuebner) [inline kong/plugins/statsd/log.lua:19]: Should be named `lmdb_get_env_info`

File: kong/plugins/statsd/log.lua
Comment (hanshuebner) [inline kong/plugins/statsd/log.lua:311]: Calculate metrics name `local metric_name = conf.hostname_in_prefix and "lmdb.used_space" or string_format("node.%s.lmdb.used_space", hostname)` or use an `if` statement to assign it, then remove the duplication of the `logger:send_statsd` lines.

**Review by hanshuebner — DISMISSED** [review]

File: .requirements
Comment (hbagdi) [inline .requirements:9]: This needs to change to a released version. Blocker for merge.

File: kong/plugins/prometheus/exporter.lua
Comment (hbagdi) [inline kong/plugins/prometheus/exporter.lua:145]:
```suggestion
    memory_stats.lmdb = prometheus:gauge("lmdb_mmap_used_bytes",
```

File: kong/plugins/prometheus/exporter.lua
Comment (hbagdi) [inline kong/plugins/prometheus/exporter.lua:149]:
```suggestion
    memory_stats.lmdb_capacity = prometheus:gauge("lmdb_mmap_size_bytes",
```

File: kong/plugins/prometheus/exporter.lua
Comment (attenuation) [reply kong/plugins/prometheus/exporter.lua:145]: I think `mmap` style metrics name is not good, Kong may have many mmap-based features in the future, but here only LMDB-related metrics are involved, so changing the name to mmap does not make sense.

Comment (chronolaw) [reply kong/plugins/prometheus/exporter.lua:145]: I agree with @attenuation , `mmap` is too technical.

File: CHANGELOG.md
Comment (chronolaw) [inline CHANGELOG.md:92]:
```suggestion
- PDK now supports getting lmdb environment information with `kong.node.get_memory_stats`.
```

File: kong/clustering/compat/init.lua
Comment (chronolaw) [inline kong/clustering/compat/init.lua:252]: Why do we use `pairs`, should it be `ipairs`?

File: kong/pdk/node.lua
Comment (chronolaw) [inline kong/pdk/node.lua:162]:
```suggestion
  --   -- if the `kong` uses dbless mode, the following will be present:
```

**Comment by chronolaw** [comment]
Please rebase this PR to the master.

File: spec/01-unit/19-hybrid/03-compat_spec.lua
Comment (chronolaw) [inline spec/01-unit/19-hybrid/03-compat_spec.lua:322]:
```suggestion
              metrics = { "lmdb_usage", "shdict_usage", "status_count_per_user_per_route", },
```

File: spec/01-unit/19-hybrid/03-compat_spec.lua
Comment (chronolaw) [inline spec/01-unit/19-hybrid/03-compat_spec.lua:330]:
```suggestion
              metrics = { "shdict_usage", "status_count_per_user_per_route", },
```

File: kong/pdk/node.lua
Comment (chronolaw) [inline kong/pdk/node.lua:131]:
```suggestion
  --    last_used_page: 6,
```

File: kong/plugins/prometheus/exporter.lua
Comment (fffonion) [reply kong/plugins/prometheus/exporter.lua:145]: let's se `lmdb_memory_used_bytes` then, having the context at front is closer to prometheus naming convention; same applies to other metrics

File: kong/plugins/prometheus/exporter.lua
Comment (oowl) [reply kong/plugins/prometheus/exporter.lua:149]: ditto

**Comment by fffonion** [comment]
code-wise lgtm, please organize this into 4 commits, PDK change, statsd plugin change, prom plugin change, changelogs

**Comment by ponch2815** [comment]
@oowl

**Comment by ponch2815** [comment]
@hbagdi @fffonion @hanshuebner

**Comment by ponch2815** [comment]
@oowl

**Comment by ponch2815** [comment]
@ponch2815 @chronolaw

**Comment by ponch2815** [comment]
@hanshuebner @ponch2815

File: kong/plugins/statsd/log.lua
Comment (dndx) [inline kong/plugins/statsd/log.lua:303]: This name seems to be misleading. `lmdb.allocated_space` maybe?

File: kong/plugins/statsd/log.lua
Comment (dndx) [inline kong/plugins/statsd/log.lua:306]: Let's just use `lmdb.map_size`, to stay consistent with the `kong.conf` entry.

File: kong/plugins/statsd/log.lua
Comment (oowl) [reply kong/plugins/statsd/log.lua:303]: offline chat whit datong for this, I will continue to contribute to this change.

File: kong/plugins/statsd/log.lua
Comment (oowl) [reply kong/plugins/statsd/log.lua:306]: ditto

**Pull request merged**
